app/models/katello/product.rb

```diff
@@ -253,5 +256,18 @@ def self.unused_product_id
     class Jail < ::Safemode::Jail
       allow :name, :label
     end
+
+    private
+
+    def ensure_provider_type_matches_id
+      if provider&.redhat_provider? && ::Katello::Glue::Candlepin::Product.custom_product_id?(cp_id)
+        errors.add(:base, _("Cannot associate a Red Hat provider with a custom product"))
+      end
+    end
+
+    def set_provider_type_from_id
+      return unless self.organization&.present? && self.organization.anonymous_provider.present? && self.organization.redhat_provider.present?
```

Comment:
When can any of these 3 entities not be present? Seems like something is really wrong in that case

Reply:
I think this was encountered in a unit test as well.

Reply:
I think the providers get created on organization create so they should be in place which brings the accuracy of those unit tests into question! of course, gotta double check what i'm saying